guppylang/definition/function.py

```diff
@@ -0,0 +1,147 @@
+import ast
+from dataclasses import dataclass, field
+
+from guppylang.ast_util import AstNode, with_loc
+from guppylang.checker.cfg_checker import CheckedCFG
+from guppylang.checker.core import Context, Globals, PyScope
+from guppylang.checker.expr_checker import check_call, synthesize_call
+from guppylang.checker.func_checker import check_global_func_def, check_signature
+from guppylang.compiler.core import CompiledGlobals, DFContainer
+from guppylang.compiler.func_compiler import compile_global_func_def
+from guppylang.definition.common import CheckableDef, CompilableDef, ParsableDef
+from guppylang.definition.value import CallableDef, CompiledCallableDef
+from guppylang.error import GuppyError
+from guppylang.hugr.hugr import DFContainingVNode, Hugr, Node, OutPortV
+from guppylang.nodes import GlobalCall
+from guppylang.tys.subst import Inst, Subst
+from guppylang.tys.ty import FunctionType, Type, type_to_row
+
+
+@dataclass(frozen=True)
+class RawFunctionDef(ParsableDef):
+    """A raw function definition provided by the user.
+
+    The raw definition stores exactly what the user has written (i.e. the AST), without
+    any additional checking or parsing. Furthermore, we store the values of the Python
+    variables in scope at the point of definition.
+    """
+
+    defined_at: ast.FunctionDef
```

Comment:
A thought: Maybe we should store a reference to the actual Python function and do the AST generation in `parse()` instead? I think this would fit better with traditional compiler lingo?

Reply:
Mhm you mean "dereferencing" in terms of looking up the source code?

The parsed `ast.FunctionDef` is an AST. There is no way to go back to a callable function from this. At the moment, we're just throwing away the Python function.

I agree, the current version works. However, I think it could be nicer to do the actual parsing in the function called `parse` instead of doing it earlier in the `@guppy` decorator?

Reply:
I implemented this in #181. I have to admit it's a bit unclean for nested function definitions. What do you think?